src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -0,0 +1,449 @@
+                {
+                    if (_currentClosure == null)
+                    {
+                        // Can't be captured if we're not in a closure
+                        return;
+                    }
+
+                    if (symbol is LocalSymbol local && local.IsConst)
+                    {
+                        // consts aren't captured since they're inlined
+                        return;
+                    }
+
+                    if (symbol.ContainingSymbol != _currentClosure.Symbol)
+                    {
+                        // Record the captured variable where it's captured
+                        var scope = _currentScope;
+                        var closure = _currentClosure;
+                        while (closure != null && symbol.ContainingSymbol != closure.Symbol)
+                        {
+                            closure.CapturedVariables.Add(symbol);
+
+                            // Also mark captured in enclosing scopes
+                            while (scope.ContainingClosure == closure)
+                            {
+                                scope = scope.Parent;
+                            }
+                            closure = scope.ContainingClosure;
+                        }
+
+                        // Also record where the captured variable lives
+
+                        // No need to record where local functions live: that was recorded
+                        // in the Closures list in each scope
+                        if (symbol.Kind == SymbolKind.Method)
+                        {
+                            return;
+                        }
+
+                        // The 'this' parameter isn't declared in method scope
+                        if (symbol is ParameterSymbol param && param.IsThis)
+                        {
+                            return;
+                        }
+
+                        if (_localToScope.TryGetValue(symbol, out var declScope))
+                        {
+                            declScope.DeclaredCapturedVariables.Add(symbol);
+                        }
+                        else
+                        {
+                            // Parameters and locals from expression tree lambdas
+                            // don't get recorded
+                            Debug.Assert(_inExpressionTree);
+                        }
+                    }
+                }
+
+                private Scope NewNestedScope(BoundNode node)
+                {
+                    // We should never create a new scope with the same bound
+                    // node. We can get into this situation for methods and
+                    // closures where a new scope is created to add parameters
+                    // and a new scope would be created for the method block,
+                    // despite the fact that they should be the same scope.
+                    if (_currentScope.BoundNode == node)
```

Comment:
The method name here speaks that it will unconditionally create a new scope. That's misleading given this behavior. Consider renaming to something along the lines of `CreateOrReuse`.